refactor: modernize code and move out helper functions

e0 opened the pull request:
**Related issue(s) and PR(s)**  
This PR closes #35.

<!-- Include below a description of the changes proposed in the pull request -->

**Type of change**  
<!-- Please delete options that are not relevant -->
- [x] Other
 
**List of changes made**  
<!-- Specify what changes have been made and why -->
- Rewrote `atlas-viewer-controls.js` as a `class`. Previously this was written as a function using things like `_this` and `prototype` to behave like a `class`. The official `three.js` has also made a similar change to their [TrackballControls.js](https://github.com/mrdoob/three.js/blob/master/examples/jsm/controls/TrackballControls.js), which our file is based on. Unlike the `TrackballControls.js` file that nests all of the code inside the `constructor`, the updated `atlas-viewer-controls.js` uses more standard practices such as class methods that hopefully makes the code more readable and easier to maintain. 
- Extracted parts from `met-atlas-viewer.js` that seems suitable to be placed into `helpers.js` The aim here was to find and move out pieces of code that do not contain key network manipulation logic, such as creating labels and materials. This should hopefully make it easier for developers to focus when working with `met-atlas-viewer.js`.
- Miscellaneous smaller changes such as replacing instances of `var` with `const/let` and using arrow functions where appropriate.


**Testing**  
<!-- Please delete options that are not relevant -->
1. Verify that the development runs: `npm start` and that the network loads as expected in the browser. Please also verify that there are no console errors.
2. Verify that the project can build without any problems: `npm run build`.

To test this out in the MetabolicAtlas repo, please refer to the following instructions.

1. Change value for dependency from version number to git branch in `frontend/package.json`.
```json
    "@metabolicatlas/3d-network-viewer": "MetabolicAtlas/3d-network-viewer#refactor/modernize-code",
```

2.  Install the dependency and build it.
```bash
ma-exec frontend yarn
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ install
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ run build
```

3. Use the code built from the previous step in `frontend/src/components/explorer/mapViewer/ThreeDviewer.vue`.
```javascript
// import { MetAtlasViewer } from '@metabolicatlas/3d-network-viewer';
import '@metabolicatlas/3d-network-viewer/public/met-atlas-viewer.js';

...

      // this.controller = MetAtlasViewer('viewer3d');
      this.controller = window.MetAtlasViewer.MetAtlasViewer('viewer3d');
```

**Further comments**  
<!-- Specify questions or related information -->
- `flatShading`, which was used for creating `PointsMaterial` instances, is no longer a valid property so it has been removed.
- The `setConnectionsColor` function in `met-atlas-viewer.js` looks attractive for refactoring since it has four segments that look almost identical with the exception of some numbers. I'm a bit worried that making this less DRY would make it more difficult to understand what's going on. Please let me know if you think this should be refactored and I'm happy to add a commit for it for us to review.
- There is currently a convention where the a list of `@authors` is at the top of some files. I removed the list from `atlas-viewer-controls.js` since it has a mention of the official `three-js` file that it's based on. This is also not a practice that we commonly have elsewhere and personally I think it's unnecessary since there is a license on this repo and you can find all of the contributors for each file in the git history. I also find it a bit tedious to have to add your name to each file after modifying it. Please let me know if you would like to keep and update these or prefer to have them removed from all the files.
- There is an `atlas-viewer-controls.js.mod` file that I don't think is necessary anymore. Could we delete this from the project?
- Please let me know if you think more things should be refactored or if any of the new additions to `helpers.js` should be moved back to `met-atlas-viewer.js`.

**Definition of Done checklist**  
- [x] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [x] My changes generate no new warnings

inghylt commented on src/helpers.js:
```diff
@@ -12,30 +19,163 @@
+ * The node should be of the shape:
+ * {id: <node ID>, pos: (<x-pos>, <y-pos>, <z-pos>), g: <group>,
+ * n: <node name>, (optional) color: [<r>, <g>, <b>],
+ *  ...
```
Perhaps silly comment, but maybe nice to add a closing bracket?

inghylt commented on src/met-atlas-viewer.js:
```diff
@@ -39,130 +40,102 @@ import { makeIndexSprite } from './helpers';
-  var scene = new Scene();
-  scene.background = new Color( 0xdddddd );
+  let scene = new Scene();
+  scene.background = new Color(0xdddddd);
```
Should this value be in helpers.js as well?

e0 replied:
Good idea. Addressed in: https://github.com/MetabolicAtlas/3d-network-viewer/pull/51/commits/f7661ce692bb7b6b94229e36185a021bd2516ac5

inghylt commented on src/met-atlas-viewer.js:
```diff
@@ -924,22 +921,22 @@ function MetAtlasViewer(targetElement) {
+  const getNodesWithin = (distance) => {
     let nodes = [];
 
     let testCamera = new PerspectiveCamera(fieldOfView, aspect, near, distance);
```
Shouldn't we be able to use helper function `moveCamera` here as well?

e0 replied:
Makes sense! I had to rewrite the `moveCamera` function a bit to accommodate the new params. Addressed in: https://github.com/MetabolicAtlas/3d-network-viewer/pull/51/commits/f7661ce692bb7b6b94229e36185a021bd2516ac5

inghylt commented:
Really nice job @e0!

FYI for the other reviewers, I had to remove node_modules to make `npm run build` work

- Regarding `setConnectionsColor`, I don't think that is very easy to understand in it's current shape either. So I'm up for trying to refactoring it and maybe make it more understandable with a comment or something.
- I like the suggestion of removing the @authors from all the files
- Yes, I think you can remove atlas-viewer-controls.js.mod

Question:
- It would be nice to test these changes out with the MetAtlas stack, but I guess there is no easy way of doings this until this is merged?

e0 commented:
Thank you for the thorough review and feedback @inghylt! I think I have implemented all the changes you suggested. Please let me know if I missed anything.

About `setConnectionsColor`, I have added some helper functions now that hopefully makes things a bit more readable. For transparency, I do not know exactly what the `modifier` from the following line does as I have only replicated the functionality prior to refactoring. I think it is just the direction but not 100% sure. If anyone would like to explore this feel free to add a comment about it.

```javascript
    mesh.geometry.attributes.color.array[index * 3 + modifier] = color[i];
```

Good idea about testing these changes before publishing a new version. I think it should be possible with [disabling the Vite pre-bundling](https://vitejs.dev/guide/dep-pre-bundling.html#browser-cache). I will try to give this a try soon.

e0 commented:
@inghylt I have updated the PR description with a section about how to test this in the MetabolicAtlas repo. Please let me know if you encounter any problems.

nanjiangshu commented on src/met-atlas-viewer.js:
```diff
@@ -496,51 +457,50 @@ function MetAtlasViewer(targetElement) {
-            x: p.x + p.x/l * cameraDistance,
-            y: p.y + p.y/l * cameraDistance,
-            z: p.z + p.z/l * cameraDistance
+            x: p.x + (p.x / l) * cameraDistance,
```
There are lots of this kind of linear mapping of vectors, maybe worth a function? Since in all use cases, the transformation matrix is a uniform scaling matrix, the function requires just one parameter.

e0 replied:
I think this is a good candidate for refactoring. I see two instances of it in the `focusOnItems` function and a few more similar ones for example in `flyUpdate`. Adding a function that works for all of these would require more parameters though. Currently I feel split on if this function should be added or if we should leave things as is since they are relatively concise but still different a bit in different instances.

nanjiangshu replied:
Fair enough. I'm OK to leave it as is.

e0 commented on src/met-atlas-viewer.js:
```diff
@@ -39,130 +43,102 @@ import { makeIndexSprite } from './helpers';
   // later
-  var nodeMesh;
-  var connectionMesh;
+  let nodeMesh;
```
I don't think this is needed. Is there any best practice about if variables should always be initialized here?

nanjiangshu replied:
It is a good practice to initialize variables as mentioned at https://www.w3schools.com/js/js_best_practices.asp
But it's up to you to implement it or not.

e0 replied:
Thanks for the link @nanjiangshu! I was reading about it on [mdn docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) and didn't see any recommendations there. I also think having `undefined` as an initial value is not helpful since it conveys nothing about the intended use so I think we can leave this as is for now.

e0 commented:
Thanks for the thorough review and feedback @nanjiangshu!

> 1. For the author, it's now showing only in the `main.js` while previously in all files. I think it is probably better to move the authors to a separate file, e.g. contributors. And we can also consider add more names to the contributors list.

Good catch! I have removed it here: https://github.com/MetabolicAtlas/3d-network-viewer/pull/51/commits/9164e2db72c0fcc0861ffa230c7665fdcee84fbf

> 2. I also think the file `atlas-viewer-controls.js.mod` should be removed. Looks like this file was created as a backup file when editing `atlas-viewer-controls.js`

This was already removed in: https://github.com/MetabolicAtlas/3d-network-viewer/pull/51/commits/fdbbbf0e4b194edc73ab5c7a7e7070a7cd76f63f.

> 3. The file `atlas-viewer-controls.js` is still a bit long, I think some functions such as `isInCamera` can be moved to helpers.js, but I'm OK to keep it as is.

Do you mean `met-atlas-viewer.js`? It does look like a good candidate for being moved out. The only change we would make is to also add `nodeInfo` as a param I think. Was there any other functions like this that you think should be moved out?

> 4. I'm OK to leave the `setConnectionsColor` un-refactored. Not quite understand why are values `[3, 4, 5]` and ` [-3, -2, -1]` set, but there are many part other functions that I don't fully understand by just looking at the code.

This was already refactored in: https://github.com/MetabolicAtlas/3d-network-viewer/pull/51/commits/d53ec1cf965fc7afe54e517316538773e8c00239.

inghylt reviewed (approved):
I've tested the changes with the rest of the MetAtlas repo and it seems to work as expected! Needed to stop and start the stack to make the instructions in the PR take affect but after that it worked. And I think the code changes are good enough for me to approve, good job!

nanjiangshu commented:
> > 3\. The file `atlas-viewer-controls.js` is still a bit long, I think some functions such as `isInCamera` can be moved to helpers.js, but I'm OK to keep it as is.
> 
> Do you mean `met-atlas-viewer.js`? It does look like a good candidate for being moved out. The only change we would make is to also add `nodeInfo` as a param I think. Was there any other functions like this that you think should be moved out?

Yes, I mean the file `met-atlas-viewer.js`. I don't know what is the best balance of size of `helpers.js` and `met-atlas-viewer.js`. In principle, all functions not listed below can be moved to `helpers.js`, but it should not be moved if it requires passing too many parameters. Again, I'm OK not moving any more functions. 
```javascript
  return {
    centerNode,
    dispose,
    setBackgroundColor,
    selectBy,
    setCameraControls,
    setColors,
    setData,
    setCamera,
    setNodeSelectCallback,
    setUpdateCameraCallback,
    setLabelDistance,
    toggleLabels,
    toggleNodeType, 
  };
```

MalinAhlberg commented:
Good job @e0! I had to restart the stack a number of times before it worked with metabolicatlas, but now is does :star:.
My experience on Chrome (Ubuntu) is that moving the network by clicking/dragging is a lot slower than it used to be. On firefox, it works as before. Just wanted to check with eg @inghylt if you experience the same?

MalinAhlberg reviewed (approved):
Everything looks good to me, and I have no strong opinions regarding the still open discussions, so I approve!
